File: app/src/main/java/uk/co/thomasrynne/swisher/sources/MediaStoreSource.scala
package uk.co.thomasrynne.swisher.sources

import android.media.{AudioManager, MediaPlayer}
import android.media.MediaPlayer.{OnPreparedListener, OnCompletionListener}
import android.net.Uri
import android.support.v4.app.Fragment
import android.util.Log
import org.json.JSONObject
import uk.co.thomasrynne.swisher.model._
import uk.co.thomasrynne.swisher.tree.{ItemMenuItem, MenuItem, Menu}
import uk.co.thomasrynne.swisher.util.{SafeMediaPlayer, MediaStore}

/**
 */
class MediaStoreSource(mediaStore:MediaStore) {

  val songsMenu = new Menu {
    def name = "Songs"
    def path = "Songs"
    def items: List[MenuItem] = songsSearch(None)
    def search: Option[(String) => List[MenuItem]] = Some( (text:String) => songsSearch(Some(text)) )
    private def songsSearch(query:Option[String]) = {
      val tracks = query match {
        case None => mediaStore.allSongs
        case Some(q) => mediaStore.searchTracks(q)
      }
      tracks.toList.map { track =>
        val json = new JSONObject
        json.put("track_name", track.name)
        json.put("artist", track.artist)
        ItemMenuItem(track.name + " (" + track.artist + ")", track.image, json, None)
      }
    }
  }

  val albumsMenu = new Menu {
    def name = "Albums"
    def path = "Albums"
    def items: List[MenuItem] = albumsSearch(None)
    def search: Option[(String) => List[MenuItem]] = Some( (text:String) => albumsSearch(Some(text)) )
    private def albumsSearch(query:Option[String]) = {
      val albums = query match {
        case None => mediaStore.allAlbums
        case Some(q) => mediaStore.searchAlbums(q)
      }
      albums.toList.map { album =>
        val json = new JSONObject
        json.put("album", album.name)
        json.put("artist", album.artist)
        ItemMenuItem(album.name + " (" + album.artist + ")", album.image, json, None)
      }
    }
  }

  val handler = new MediaHandler {
    def handle(json: JSONObject) = {
      if (json.has("artist") && json.has("album")) {
        val album = json.getString("album")
        val artist = json.getString("artist")
        mediaStore.albumAndTracks(album, artist) match {
          case None => NotAvailableHandleResult
          case Some((album,tracks)) => SuccessHandleResult(new MediaPlayerAlbumTracks(album, tracks.toList))
        }
      } else if (json.has("track_name") && json.has("artist")) {
        val trackName = json.getString("track_name")
        val artist = json.getString("artist")
        Option(mediaStore.findTrackOrNull(trackName, artist)) match {
          case None => NotAvailableHandleResult
          case Some(track) => SuccessHandleResult(new MediaPlayerSingleTrackTracks(track))
        }
      } else {
        NotApplicableHandleResult
      }
    }
  }

  class MediaPlayerAlbumTracks(album:mediaStore.Album, tracksX:List[mediaStore.Track]) extends Tracks {
    def tracks:List[TrackDescription] = tracksX
    def name = album.name
    def thumbnail: Option[Uri] = album.image
    def player(init:PlayerInit, listener: PlayerListener) = new MediaPlayerTracksPlayer(tracksX.toArray, init, listener)
  }
  class MediaPlayerTracksPlayer(tracks:Array[mediaStore.Track],
                                init:PlayerInit, listener:PlayerListener) extends TracksPlayer {
    private var current = 0
    private val mediaPlayer = new SafeMediaPlayer(
      onComplete = () => {
        Log.i("SWISHER", "onComplete " + current)
        if ((current + 1) < tracks.length) {
          current += 1
          playCurrent()
          listener.onTrack(current)
        } else {
          listener.finished()
        }
      },
      onProgressChange = (progress:PlayerProgress) =>
        listener.updateProgress(progress)
    )
    init match {
      case CuePlayerInit => mediaPlayer.play(tracks(0).path, 0, false); listener.onTrack(0)
      case PlayTrackInit(track) => jumpTo(track)
      case PlayPlayerInit => mediaPlayer.play(tracks(0).path, 0, true); listener.onTrack(0)
    }

    private def playCurrent() { mediaPlayer.play(tracks(current).path) }
    override def progress = mediaPlayer.progress
    def cue() { current = 0; mediaPlayer.play(tracks(0).path, 0, false); listener.onTrack(0) }
    def pausePlay() { mediaPlayer.pausePlay }
    def clear() { mediaPlayer.release }
    def stop() { mediaPlayer.stop }
    def jumpTo(trackIndex: Int) { current=trackIndex; playCurrent() }
    def seekTo(millis: Int) { mediaPlayer.seekTo(millis) }
  }

  class MediaPlayerSingleTrackTracks(track:mediaStore.Track) extends Tracks {
    def name = track.name
    def tracks: List[TrackDescription] = List(track)
    def thumbnail: Option[Uri] = track.image
    def player(init:PlayerInit, listener: PlayerListener) = {
      new MediaPlayerTracksPlayer(Array(track), init, listener)
    }
  }
}

File: app/src/main/java/uk/co/thomasrynne/swisher/sources/MediaStoreSource.java
package uk.co.thomasrynne.swisher.sources;

import android.content.Context;
import android.net.Uri;
import android.provider.BaseColumns;
import android.provider.MediaStore;

import com.getbase.android.db.cursors.FluentCursor;
import com.getbase.android.db.provider.ProviderAction;
import com.google.common.base.Optional;
import com.google.common.collect.FluentIterable;
import com.google.common.collect.ImmutableList;

import java.util.Collections;
import java.util.List;

import lombok.Value;
import lombok.val;
import thomas.swisher.MediaHandler;
import thomas.swisher.localmedia.AsyncMediaPlayer;
import thomas.swisher.tree.MenuUtils;
import thomas.swisher.tree.Menus;
import uk.co.thomasrynne.swisher.Songs;
import uk.co.thomasrynne.swisher.Utils;
import thomas.swisher.shared.Core;
import thomas.swisher.tree.MainMenuTree;
import uk.co.thomasrynne.swisher.core.TracksPlayer;

/**
 */
public class MediaStoreSource {

    private Context context;

    public MediaStoreSource(Context context) {
        this.context = context;
    }

    public Menus.Menu albumMenu() {
        return new Menus.Menu() {
            @Override
            public String name() {
                return "Albums";
            }

            @Override
            public List<Menus.MenuEntry> items() {
                return MenuUtils.alphabet(albums());
            }
        };
    }

    public Menus.Menu tracksMenu() {
        return new Menus.Menu() {
            @Override
            public String name() {
                return "Tracks";
            }

            @Override
            public List<Menus.MenuEntry> items() {
                return MenuUtils.alphabet(tracks());
            }
        };
    }

    public MediaHandler albumHandler() {
        return new MediaHandler() {
            @Override
            public Optional<PlaylistEntry> handle(Utils.FlatJson json) {
                if (json.has("artist") && json.has("album")) {
                    val album = json.get("album");
                    val artist = json.get("artist");
                    val albumId = albumIdFor(album, artist);
                    val name = album + " (" + artist + ")";
                    if (albumId.isPresent()) {
                        val tracks = albumTracks(albumId.get());
                        return Optional.of(new MediaHandler.PlaylistEntry(
                                name,
                                Songs.imageForAlbum(albumId.get()),
                                tracks,
                                (playNow, playNext, listener) -> new LocalMediaPlayer(tracks, playNow, playNext, listener)
                        ));
                    } else {
                        return Optional.absent();
                    }
                } else {
                    return Optional.absent();
                }
            }
        };
    }


    public MediaHandler trackHandler() {
        return new MediaHandler() {
            @Override
            public Optional<PlaylistEntry> handle(Utils.FlatJson json) {
                if (json.has("track_name") && json.has("artist")) {
                    val trackName = json.get("track_name");
                    val artist = json.get("artist");
                    val maybeTrack = findTrack(trackName, artist);
                    if (maybeTrack.isPresent()) {
                        val track = maybeTrack.get();
                        val singleTrack = Collections.singletonList(track);
                        return Optional.of(new MediaHandler.PlaylistEntry(
                                track.name(),
                                track.image,
                                singleTrack,
                                (playNow, playNext, listener) -> new LocalMediaPlayer(singleTrack, playNow, playNext, listener)
                        ));
                    } else {
                        return Optional.absent();
                    }
                } else {
                    return Optional.absent();
                }
            }
        };
    }

    private List<Core.PlaylistItemUIMenuItem> albums() {
        val action = ProviderAction.query(MediaStore.Audio.Albums.EXTERNAL_CONTENT_URI)
            .projection(BaseColumns._ID, MediaStore.Audio.AlbumColumns.ALBUM, MediaStore.Audio.AudioColumns.ARTIST)
            .orderBy(MediaStore.Audio.AudioColumns.ALBUM)
            .perform(context.getContentResolver());
        return action.toFluentIterable( (cursor) -> {

            int albumIndex = cursor.getColumnIndexOrThrow(MediaStore.Audio.AudioColumns.ALBUM);
            int artistIndex = cursor.getColumnIndex(MediaStore.Audio.AudioColumns.ARTIST);
            int albumIdIndex = cursor.getColumnIndexOrThrow(BaseColumns._ID);

            String albumName = cursor.getString(albumIndex);
            String artistName = cursor.getString(artistIndex);
            int albumId = cursor.getInt(albumIdIndex);
            return new Core.PlaylistItemUIMenuItem(
                    cursor.getString(albumIndex),
                    Utils.json().add("album", albumName).add("artist", artistName).build(),
                    Songs.imageForAlbum(albumId));
        }).toList();
    }

    private Optional<Integer> albumIdFor(String album, String artist) {
        val selection = MediaStore.Audio.AudioColumns.ALBUM + "=? " + " and " + MediaStore.Audio.AudioColumns.ARTIST + "=?";
        val action = ProviderAction.query(MediaStore.Audio.Albums.EXTERNAL_CONTENT_URI)
                .projection(BaseColumns._ID, MediaStore.Audio.AudioColumns.ALBUM, MediaStore.Audio.AudioColumns.ARTIST)
                .where(selection, album, artist)
                .perform(context.getContentResolver());
        return action.toFluentIterable( (cursor) -> {
            int albumIdIndex = cursor.getColumnIndexOrThrow(BaseColumns._ID);
            return cursor.getInt(albumIdIndex);
        }).first();
    }

    @Value
    static class Track implements MediaHandler.TrackDescription {
        public final String trackTitle;
        public final String artistName;
        public final int track;
        public final Uri path;
        public final Optional<Uri> image;

        @Override
        public String name() {
            return trackTitle;
        }

        @Override
        public Optional<Uri> image() {
            return image;
        }
    }

    private ImmutableList<Track> albumTracks(int albumId) {
        val action = ProviderAction.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI)
                .projection(
                        MediaStore.MediaColumns.TITLE,
                        MediaStore.Audio.AudioColumns.ARTIST,
                        MediaStore.Audio.AudioColumns.TRACK,
                        MediaStore.Audio.AlbumColumns.ALBUM_ID,
                        MediaStore.MediaColumns.DATA)
                .where(MediaStore.Audio.AlbumColumns.ALBUM_ID + " = ?", albumId)
                .orderBy(android.provider.MediaStore.Audio.AudioColumns.TRACK)
                .perform(context.getContentResolver());
        return FluentIterable.from(tracksFromAction(action).toSortedSet( (a,b) -> a.track - b.track)).toList();
    }

    private Optional<Track> findTrack(String trackName, String artist) {
        val action = ProviderAction.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI)
                .projection(
                        MediaStore.MediaColumns.TITLE,
                        MediaStore.Audio.AudioColumns.ARTIST,
                        MediaStore.Audio.AudioColumns.TRACK,
                        MediaStore.Audio.AlbumColumns.ALBUM_ID,
                        MediaStore.MediaColumns.DATA)
                .where(MediaStore.MediaColumns.TITLE + " =? and " + MediaStore.Audio.AudioColumns.ARTIST + " =?",
                        trackName, artist
                )
                .perform(context.getContentResolver());
        return tracksFromAction(action).first();
    }

    private FluentIterable<Track> tracksFromAction(FluentCursor action) {
        return action.toFluentIterable((cursor) -> {
            int titleIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.TITLE);
            int artistIndex = cursor.getColumnIndex(MediaStore.Audio.AudioColumns.ARTIST);
            int trackIndex = cursor.getColumnIndex(MediaStore.Audio.AudioColumns.TRACK);
            int albumIdIndex = cursor.getColumnIndex(MediaStore.Audio.AlbumColumns.ALBUM_ID);
            int dataIndex = cursor.getColumnIndex(MediaStore.Audio.AudioColumns.DATA);

            String title = cursor.getString(titleIndex);
            String artistName = cursor.getString(artistIndex);
            int index = cursor.getInt(trackIndex);
            int albumId = cursor.getInt(albumIdIndex);
            String path = cursor.getString(dataIndex);

            return new Track(
                    title,
                    artistName,
                    index,
                    Uri.parse(path),
                    Songs.imageForAlbum(albumId));
        });
    }

    private ImmutableList<Core.PlaylistItemUIMenuItem> tracks() {
        val action = ProviderAction.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI)
                .projection(
                        MediaStore.MediaColumns.TITLE,
                        MediaStore.Audio.AudioColumns.ARTIST,
                        MediaStore.Audio.AlbumColumns.ALBUM_ID)
                .orderBy(MediaStore.MediaColumns.TITLE)
                .perform(context.getContentResolver());
        return action.toFluentIterable((cursor) -> {

            int titleIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.TITLE);
            int artistIndex = cursor.getColumnIndex(MediaStore.Audio.AudioColumns.ARTIST);
            int albumIdIndex = cursor.getColumnIndex(MediaStore.Audio.AlbumColumns.ALBUM_ID);

            String title = cursor.getString(titleIndex);
            String artistName = cursor.getString(artistIndex);
            int albumId = cursor.getInt(albumIdIndex);
            String name = title + " (" + artistName + ")";

            return new Core.PlaylistItemUIMenuItem(
                    name,
                    Utils.json().add("track_name", title).add("artist", artistName).build(),
                    Songs.imageForAlbum(albumId));
        }).toList();
    }

    private class LocalMediaPlayer implements TracksPlayer {
        private List<Track> tracks;
        private MediaHandler.PlayerListener listener;
        private int currentTrack = 0;
        private boolean playNext;
        private AsyncMediaPlayer mediaPlayer = new AsyncMediaPlayer(
                () -> {
                    if ((currentTrack + 1) < tracks.size()) {
                        currentTrack++;
                        playCurrent(playNext);
                        listener.onTrack(currentTrack, playNext);
                    } else {
                        listener.finished();
                    }
                }
        );

        private void playCurrent(boolean playNow) {
            mediaPlayer.play(tracks.get(currentTrack).path.getPath(), 0, playNow);
        }

        LocalMediaPlayer(List<Track> tracks, boolean playNow, boolean playNext, MediaHandler.PlayerListener listener) {
            this.tracks = tracks;
            this.playNext = playNext;
            this.listener = listener;
            playCurrent(playNow);
        }

        @Override
        public void cueBeginning() {
            currentTrack = 0;
            playCurrent(false);
        }

        @Override
        public void stop() {
            mediaPlayer.stop();
        }

        @Override
        public void pausePlay() {
            mediaPlayer.pausePlay();
        }

        @Override
        public void clear() {
            mediaPlayer.release();
        }

        @Override
        public void playNext(boolean playNext) {
            this.playNext = playNext;
        }

        @Override
        public void jumpTo(int track) {
            currentTrack=track;
            playCurrent(true);
        }
    }
}
